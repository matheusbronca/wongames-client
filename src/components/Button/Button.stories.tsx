import { Meta, StoryObj } from '@storybook/react'
import Button from './Button'
import { ButtonProps } from './Button'
import Icon from '@svg-icons/material-outlined/add-shopping-cart.svg'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    isFullWidth: {
      type: 'boolean'
    },
    Icon: {
      type: 'string'
    }
  }
} as Meta

export const Default: StoryObj<ButtonProps> = {
  render: (args: ButtonProps) => <Button {...args}>{args.children}</Button>
}

Default.args = {
  children: 'Buy now'
}

export const WithIcon: StoryObj<ButtonProps> = {
  render: (args: ButtonProps) => <Button {...args}>{args.children}</Button>
}

WithIcon.args = {
  children: 'Buy now',
  Icon: <Icon />
}

export const AsLink: StoryObj<ButtonProps> = {
  render: (args: ButtonProps) => (
    <Button role="a" {...args}>
      {args.children}
    </Button>
  )
}

AsLink.args = {
  children: 'Buy now'
}
