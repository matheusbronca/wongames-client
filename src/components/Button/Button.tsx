import { cn } from '@/lib/utils'
import { cva } from 'cva'
import { AnchorHTMLAttributes, Attributes, ButtonHTMLAttributes } from 'react'
import React from 'react'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  isFullWidth?: boolean
  Icon?: JSX.Element
  renderAs?: Extract<React.ElementType, 'button' | 'a'>
} & ButtonTypes

const getVariantClasses = cva(null, {
  variants: {
    size: {
      small: 'text-xs h-8 p-2',
      medium: 'text-sm  h-10 px-md',
      large: 'text-base h-12 px-xlg'
    }
  }
})

const Button = ({
  children,
  size = 'medium',
  isFullWidth = false,
  Icon,
  renderAs = 'button',
  ...props
}: ButtonProps) => {
  // Needed to create tag "on demand"
  const Tag = ({ renderAs, children, ...props }: ButtonProps) =>
    React.createElement(
      renderAs as Extract<ButtonProps, 'renderAs'>,
      props as Attributes,
      children
    )

  return (
    <Tag
      renderAs={renderAs}
      className={cn(
        'inline-flex items-center justify-center text-white no-underline border-none rounded bg-gradient-to-b cursor-pointer from-[#FF5F5F] to-[#F062C0] hover:bg-gradient-to-b hover:from-[#e35565] hover:to-[#d958a6]',
        getVariantClasses({
          size
        }),
        {
          'w-full': Boolean(isFullWidth)
        },
        {
          'inline-flex items-center justify-center [&_>_svg]:w-4 [&_svg_+_span]:ml-xxsm':
            !!Icon
        }
      )}
      {...props}
    >
      {!!Icon && Icon}
      {!!children && <span>{children}</span>}
    </Tag>
  )
}

export default Button
