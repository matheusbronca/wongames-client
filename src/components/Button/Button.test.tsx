import { render, screen } from '@testing-library/react'
import { twMerge } from 'tailwind-merge'
import ShoppingCartIcon from '@svg-icons/material-outlined/shopping-cart.svg'

import Button from './Button'

describe('<Button />', () => {
  const COMMON_BTN = 'border-none rounded'
  const SM_BTN = 'text-xs h-7 p-2'
  const MD_BTN = 'text-sm  h-10 px-md'
  const LG_BTN = 'text-base h-12 px-xlg'

  it('should render the Button with medium size by default', () => {
    render(<Button>Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveClass(
      twMerge([COMMON_BTN, MD_BTN])
    )

    expect(screen.getByRole('button', { name: /Buy now/i })).toMatchSnapshot()
  })

  it('should render the Button with small size', () => {
    render(<Button size="small">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveClass(
      twMerge([COMMON_BTN, SM_BTN])
    )
  })

  it('should render the Button with large size', () => {
    render(<Button size="large">Buy now</Button>)

    expect(screen.getByRole('button', { name: /Buy now/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveClass(
      twMerge([COMMON_BTN, LG_BTN])
    )
  })

  it('should render the Button with full width', () => {
    render(
      <>
        <Button size="large" isFullWidth={true}>
          Large
        </Button>
        <Button size="medium" isFullWidth={true}>
          Medium
        </Button>
        <Button size="small" isFullWidth={true}>
          Small
        </Button>
      </>
    )

    expect(screen.getByRole('button', { name: /Large/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Large/i })).toHaveClass(
      twMerge([COMMON_BTN, LG_BTN, 'w-full'])
    )

    expect(screen.getByRole('button', { name: /Medium/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Medium/i })).toHaveClass(
      twMerge([COMMON_BTN, MD_BTN, 'w-full'])
    )

    expect(screen.getByRole('button', { name: /Small/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /Small/i })).toHaveClass(
      twMerge([COMMON_BTN, SM_BTN, 'w-full'])
    )
  })

  it('should render the button with an icon when provided via componet props', () => {
    render(
      <Button Icon={<ShoppingCartIcon data-testid="svg" />}>Buy now</Button>
    )
    expect(screen.getByTestId('svg')).toBeInTheDocument()
  })

  it('should render Button as an anchor link', () => {
    render(
      <Button renderAs="a" href="/link">
        Anchor Link
      </Button>
    )

    expect(screen.getByRole('link', { name: /anchor link/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
