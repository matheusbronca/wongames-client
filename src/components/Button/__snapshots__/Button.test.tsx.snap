// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the Button with medium size by default 1`] = `
<button
  class="inline-flex items-center justify-center text-white no-underline border-none rounded bg-gradient-to-b cursor-pointer from-[#FF5F5F] to-[#F062C0] hover:bg-gradient-to-b hover:from-[#e35565] hover:to-[#d958a6] text-sm h-10 px-md"
>
  <span>
    Buy now
  </span>
</button>
`;
